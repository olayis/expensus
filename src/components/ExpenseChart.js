import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChartBar } from '@fortawesome/free-regular-svg-icons';
import charts from '../../public/img/charts.svg';
import categoryCountAmountPercentage from '../statistics/categoryCountAmount';
import dailyAmount from '../statistics/dailyAmount';
import dailyCount from '../statistics/dailyCount';
import expenseAmount from '../statistics/descriptionAmount';
import ExpenseChartItem from './ExpenseChartItem';
import NoExpense from './NoExpense';

const ExpenseChart = (props) => {
  const pieChartData = expenseAmount(props.expenses);
  const barChartData = dailyAmount(props.expenses);
  const lineChartData = dailyCount(props.expenses);
  const bubbleChartData = categoryCountAmountPercentage(props.expenses);

  return (
    <section className='content-container'>
      <div className='component-container component-container--chart'>
        <h2 className='heading-secondary'>
          <FontAwesomeIcon icon={faChartBar} /> Expenses Charts
        </h2>
        {props.expenses.length === 0 ? (
          <NoExpense
            expensesWithoutFilters={props.expensesWithoutFilters}
            noDataImage={charts}
            noDataImageAlt='A man looking at charts'
            filterImage={charts}
            filterImageAlt='A man looking at charts'
            noExpenseText='Start adding expenses to generate charts.'
          />
        ) : (
          <div>
            <div className='row'>
              <div className='col-1-of-2'>
                <ExpenseChartItem
                  title='Expenses (By Percentage)'
                  width='100%'
                  height='30rem'
                  chartType='PieChart'
                  data={pieChartData}
                  chartRootProps={{ 'data-testid': '1' }}
                />
              </div>
              <div className='col-1-of-2'>
                <ExpenseChartItem
                  title='Daily Amount Of Expenses'
                  width='100%'
                  height='30rem'
                  chartType='Bar'
                  data={barChartData}
                  chartRootProps={{ 'data-testid': '2' }}
                />
              </div>
            </div>
            <div className='row'>
              <ExpenseChartItem
                title='Daily Number of Expenses'
                width='100%'
                height='30rem'
                chartType='LineChart'
                data={lineChartData}
                chartRootProps={{ 'data-testid': '3' }}
                options={{
                  hAxis: {
                    title: 'Date',
                  },
                  vAxis: {
                    title: 'Number',
                  },
                }}
              />
            </div>
            <div className='row'>
              <ExpenseChartItem
                title='Daily Number of Expenses'
                width='100%'
                height='50rem'
                chartType='BubbleChart'
                data={bubbleChartData}
                chartRootProps={{ 'data-testid': '4' }}
                options={{
                  hAxis: { title: 'Category' },
                  vAxis: { title: 'Number' },
                }}
              />
            </div>
          </div>
        )}
      </div>
    </section>
  );
};

export default ExpenseChart;
