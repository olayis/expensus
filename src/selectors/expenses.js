// Get Visible expenses
import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      switch (sortBy) {
        case 'date(recent)':
          return a.createdAt < b.createdAt ? 1 : -1;

        case 'date(oldest)':
          return a.createdAt > b.createdAt ? 1 : -1;

        case 'amount(highest)':
          return a.amount < b.amount ? 1 : -1;

        case 'amount(lowest)':
          return a.amount > b.amount ? 1 : -1;

        case 'description(a_to_z)':
          return a.description.localeCompare(b.description);

        case 'description(z_to_a)':
          return b.description.localeCompare(a.description);

        case 'category':
          return a.category.localeCompare(b.category);
      }
    });
};
