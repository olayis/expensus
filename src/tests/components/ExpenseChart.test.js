import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpenseChart from '../../components/ExpenseChart';

test('should render ExpensesChart with expenses data', () => {
  const wrapper = shallow(<ExpenseChart expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render NoExpense component if there are no visible expenses', () => {
  const wrapper = shallow(<ExpenseChart expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
