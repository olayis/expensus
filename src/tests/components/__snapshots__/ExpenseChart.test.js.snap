// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesChart with expenses data 1`] = `
<section
  className="content-container"
>
  <div
    className="component-container component-container--chart"
  >
    <h2
      className="heading-secondary"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f080",
              "M396.8 352h22.4c6.4 0 12.8-6.4 12.8-12.8V108.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v230.4c0 6.4 6.4 12.8 12.8 12.8zm-192 0h22.4c6.4 0 12.8-6.4 12.8-12.8V140.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v198.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h22.4c6.4 0 12.8-6.4 12.8-12.8V204.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v134.4c0 6.4 6.4 12.8 12.8 12.8zM496 400H48V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v336c0 17.67 14.33 32 32 32h464c8.84 0 16-7.16 16-16v-16c0-8.84-7.16-16-16-16zm-387.2-48h22.4c6.4 0 12.8-6.4 12.8-12.8v-70.4c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v70.4c0 6.4 6.4 12.8 12.8 12.8z",
            ],
            "iconName": "chart-bar",
            "prefix": "far",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
       Expenses Charts
    </h2>
    <div>
      <div
        className="row"
      >
        <div
          className="col-1-of-2"
        >
          <ExpenseChartItem
            chartRootProps={
              Object {
                "data-testid": "1",
              }
            }
            chartType="PieChart"
            data={
              Array [
                Array [
                  "Title",
                  "Amount",
                ],
                Array [
                  "Data",
                  20,
                ],
                Array [
                  "Home Rentage",
                  350,
                ],
                Array [
                  "Room Repainting",
                  80,
                ],
                Array [
                  "Water Pipes Repair",
                  10,
                ],
              ]
            }
            height="30rem"
            title="Expenses (By Percentage)"
            width="100%"
          />
        </div>
        <div
          className="col-1-of-2"
        >
          <ExpenseChartItem
            chartRootProps={
              Object {
                "data-testid": "2",
              }
            }
            chartType="Bar"
            data={
              Array [
                Array [
                  "Date",
                  "Amount",
                ],
                Array [
                  "01/01",
                  20,
                ],
                Array [
                  "27/12",
                  350,
                ],
                Array [
                  "13/01",
                  80,
                ],
                Array [
                  "11/01",
                  10,
                ],
              ]
            }
            height="30rem"
            title="Daily Amount Of Expenses"
            width="100%"
          />
        </div>
      </div>
      <div
        className="row"
      >
        <ExpenseChartItem
          chartRootProps={
            Object {
              "data-testid": "3",
            }
          }
          chartType="LineChart"
          data={
            Array [
              Array [
                "Date",
                "Number Of Expenses",
              ],
              Array [
                "01/01",
                1,
              ],
              Array [
                "27/12",
                1,
              ],
              Array [
                "13/01",
                1,
              ],
              Array [
                "11/01",
                1,
              ],
            ]
          }
          height="30rem"
          options={
            Object {
              "hAxis": Object {
                "title": "Date",
              },
              "vAxis": Object {
                "title": "Number",
              },
            }
          }
          title="Daily Number of Expenses"
          width="100%"
        />
      </div>
      <div
        className="row"
      >
        <ExpenseChartItem
          chartRootProps={
            Object {
              "data-testid": "4",
            }
          }
          chartType="BubbleChart"
          data={
            Array [
              Array [
                "ID",
                "Total",
                "Number",
                "Category",
                "Percentage",
              ],
              Array [
                "Bil...",
                20,
                1,
                "Bills & Utilities",
                4.3478260869565215,
              ],
              Array [
                "Ren...",
                350,
                1,
                "Rent",
                76.08695652173914,
              ],
              Array [
                "Mai...",
                90,
                2,
                "Maintenance",
                19.565217391304348,
              ],
            ]
          }
          height="50rem"
          options={
            Object {
              "hAxis": Object {
                "title": "Category",
              },
              "vAxis": Object {
                "title": "Number",
              },
            }
          }
          title="Daily Number of Expenses"
          width="100%"
        />
      </div>
    </div>
  </div>
</section>
`;

exports[`should render NoExpense component if there are no visible expenses 1`] = `
<section
  className="content-container"
>
  <div
    className="component-container component-container--chart"
  >
    <h2
      className="heading-secondary"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f080",
              "M396.8 352h22.4c6.4 0 12.8-6.4 12.8-12.8V108.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v230.4c0 6.4 6.4 12.8 12.8 12.8zm-192 0h22.4c6.4 0 12.8-6.4 12.8-12.8V140.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v198.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h22.4c6.4 0 12.8-6.4 12.8-12.8V204.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v134.4c0 6.4 6.4 12.8 12.8 12.8zM496 400H48V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v336c0 17.67 14.33 32 32 32h464c8.84 0 16-7.16 16-16v-16c0-8.84-7.16-16-16-16zm-387.2-48h22.4c6.4 0 12.8-6.4 12.8-12.8v-70.4c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v70.4c0 6.4 6.4 12.8 12.8 12.8z",
            ],
            "iconName": "chart-bar",
            "prefix": "far",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
       Expenses Charts
    </h2>
    <Connect(NoExpense)
      filterImage="SvgrURL"
      filterImageAlt="A man looking at charts"
      noDataImage="SvgrURL"
      noDataImageAlt="A man looking at charts"
      noExpenseText="Start adding expenses to generate charts."
    />
  </div>
</section>
`;
